tool/tsh/common/app.go

```diff
@@ -534,126 +486,190 @@ func serializeAppConfig(configInfo *appConfigInfo, format string) (string, error
-	caTLSCert, err := keys.LoadX509KeyPair(caPath, keyPath)
-	if err == nil {
-		if expire, err := getTLSCertExpireTime(caTLSCert); err == nil && time.Now().Before(expire) {
-			return caTLSCert, nil
+// checkAndSetDefaults checks the app route, applies cli flags, and sets defaults.
+func (a *appInfo) checkAndSetDefaults(cf *CLIConf, tc *client.TeleportClient, profile *client.ProfileStatus) error {
+	if a.IsAWSConsole() {
+		app, err := a.GetApp(cf.Context, tc)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		awsRoleARN, err := getARNFromFlags(cf, profile, app)
+		if err != nil {
+			return trace.Wrap(err)
 		}
+		a.AWSRoleARN = awsRoleARN
 	}
-	if err != nil && !trace.IsNotFound(err) {
-		log.WithError(err).Debugf("Failed to load certificate from %v.", caPath)
+
+	if a.IsAzureCloud() {
+		azureIdentity, err := getAzureIdentityFromFlags(cf, profile)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		log.Debugf("Azure identity is %q", azureIdentity)
+		a.AzureIdentity = azureIdentity
 	}
 
-	// Generate and load again.
-	if err = generateSelfSignedCA(caPath, keyPath, validUntil, dnsNames...); err != nil {
-		return tls.Certificate{}, err
+	if a.IsGCP() {
+		gcpServiceAccount, err := getGCPServiceAccountFromFlags(cf, profile)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		log.Debugf("GCP service account is %q", gcpServiceAccount)
+		a.GCPServiceAccount = gcpServiceAccount
 	}
 
-	caTLSCert, err = keys.LoadX509KeyPair(caPath, keyPath)
-	if err != nil {
-		return tls.Certificate{}, trace.Wrap(err)
+	return nil
+}
+
+func (a *appInfo) IsAWSConsole() bool {
+	if a.app != nil {
+		return a.app.IsAWSConsole()
 	}
-	return caTLSCert, nil
+	return a.RouteToApp.AWSRoleARN != ""
 }
 
-// generateSelfSignedCA generates a new self-signed CA for provided dnsNames
-// and saves/overwrites the local CA file in the profile directory.
-func generateSelfSignedCA(caPath, keyPath string, validUntil time.Time, dnsNames ...string) error {
-	log.Debugf("Generating local self signed CA at %v", caPath)
-	keyPem, err := utils.ReadPath(keyPath)
-	if err != nil {
-		return trace.Wrap(err)
+func (a *appInfo) IsAzureCloud() bool {
+	if a.app != nil {
+		return a.app.IsAzureCloud()
+	}
+	return a.RouteToApp.AzureIdentity != ""
+}
+
+func (a *appInfo) IsGCP() bool {
+	if a.app != nil {
+		return a.app.IsGCP()
 	}
+	return a.RouteToApp.GCPServiceAccount != ""
+}
 
-	key, err := keys.ParsePrivateKey(keyPem)
+// GetApp returns the cached app or fetches it using the app route and
+// caches the result.
+func (a *appInfo) GetApp(ctx context.Context, tc *client.TeleportClient) (types.Application, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	if a.app != nil {
+		return a.app, nil
+	}
+	// holding mutex across the api call to avoid multiple redundant api calls.
+	app, err := getApp(ctx, tc, a.Name)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
+	a.app = app
+	return a.app, nil
```

Comment:
to make using this safer, we should probably return `a.app.Copy()` here (and same for `databaseInfo.GetDatabase` which this was based on)